rust/arrow/src/memory.rs

```diff
@@ -38,11 +38,19 @@ pub fn free_aligned(p: *mut u8, size: usize) {
 
 pub fn reallocate(ptr: *mut u8, old_size: usize, new_size: usize) -> *mut u8 {
     unsafe {
-        std::alloc::realloc(
+        let new_ptr = std::alloc::realloc(
             ptr,
             Layout::from_size_align_unchecked(old_size, ALIGNMENT),
             new_size,
-        )
+        );
+        if new_size > old_size {
```

Comment:
There should be an additional check here in case `realloc` [returns a null pointer (indicating an error)](https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#method.realloc), since otherwise the [`offset` will be out-of-bounds](https://doc.rust-lang.org/std/primitive.pointer.html#safety-1). (Also, `std::ptr::write_bytes` [requires the pointer to be non-null](https://doc.rust-lang.org/std/ptr/fn.write_bytes.html#safety).)
```suggestion
        if !new_ptr.is_null() && new_size > old_size {
```